packages/fresh/src/middlewares/ip_restriction.ts

```diff
@@ -0,0 +1,116 @@
+import type { Context } from "../context.ts";
+import type { Middleware } from "./mod.ts";
+import { isIPv4, isIPv6, matchSubnets } from "@std/net/unstable-ip";
+
+export type AddressType = "IPv6" | "IPv4" | "none";
+export interface IPRestrictionRules {
+  denyList?: string[];
+  allowList?: string[];
+}
+
+function buildMatcher(ipList: string[]) {
+  return (addr: string) => {
+    return matchSubnets(addr, ipList);
+  };
+}
+
+function blockError(): Response {
+  return new Response("Forbidden", {
+    status: 403,
+    headers: {
+      "Content-Type": "text/plain",
+    },
+  });
+}
+
+function findType(addr: string): AddressType {
+  if (isIPv4(addr)) {
+    return "IPv4";
+  } else if (isIPv6(addr)) {
+    return "IPv6";
+  }
+  return "none";
+}
+
+export interface IpRestrictionOptions {
+  onError?: <State>(
+    remote: { addr: string; type: AddressType },
+    ctx: Context<State>,
+  ) => Response | Promise<Response>;
+}
+
+/**
+ * IP Restriction Middleware for Fresh.
```

Comment:
Nit. Also, is that a sufficient description of this middleware?
```suggestion
 * IP restriction middleware for Fresh.
```